source/credentials_provider_ecs.c

```diff
@@ -0,0 +1,636 @@
+        impl->function_table->aws_http_connection_close(ecs_user_data->connection);
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) ECS credentials provider query response exceeded maximum allowed length",
+            (void *)ecs_user_data->ecs_provider);
+
+        return AWS_OP_ERR;
+    }
+
+    if (aws_byte_buf_append_dynamic(&ecs_user_data->current_result, data)) {
+        impl->function_table->aws_http_connection_close(ecs_user_data->connection);
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) ECS credentials provider query error appending response",
+            (void *)ecs_user_data->ecs_provider);
+
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_ecs_on_incoming_headers_fn(
+    struct aws_http_stream *stream,
+    enum aws_http_header_block header_block,
+    const struct aws_http_header *header_array,
+    size_t num_headers,
+    void *user_data) {
+
+    (void)header_array;
+    (void)num_headers;
+
+    if (header_block != AWS_HTTP_HEADER_BLOCK_MAIN) {
+        return AWS_OP_SUCCESS;
+    }
+
+    struct aws_credentials_provider_ecs_user_data *ecs_user_data = user_data;
+    if (header_block == AWS_HTTP_HEADER_BLOCK_MAIN) {
+        if (ecs_user_data->status_code == 0) {
+            struct aws_credentials_provider_ecs_impl *impl = ecs_user_data->ecs_provider->impl;
+            if (impl->function_table->aws_http_stream_get_incoming_response_status(
+                    stream, &ecs_user_data->status_code)) {
+
+                AWS_LOGF_ERROR(
+                    AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                    "(id=%p) ECS credentials provider failed to get http status code",
+                    (void *)ecs_user_data->ecs_provider);
+
+                return AWS_OP_ERR;
+            }
+            AWS_LOGF_DEBUG(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p) ECS credentials provider query received http status code %d",
+                (void *)ecs_user_data->ecs_provider,
+                ecs_user_data->status_code);
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_ecs_query_task_role_credentials(struct aws_credentials_provider_ecs_user_data *ecs_user_data);
+
+static void s_ecs_on_stream_complete_fn(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct aws_credentials_provider_ecs_user_data *ecs_user_data = user_data;
+
+    (void)error_code;
+
+    aws_http_message_destroy(ecs_user_data->request);
+    ecs_user_data->request = NULL;
+
+    struct aws_credentials_provider_ecs_impl *impl = ecs_user_data->ecs_provider->impl;
+    impl->function_table->aws_http_stream_release(stream);
+
+    /*
+     * On anything other than a 200, nullify the response and pretend there was
+     * an error
+     */
+    if (ecs_user_data->status_code != 200) {
+        ecs_user_data->current_result.len = 0;
+    }
```

Comment:
probably also want to nullify the answer if `error_code` is non-zero. Currently `error_code` is ignored